bindings/matrix-sdk-ffi/src/spaces.rs

```diff
@@ -0,0 +1,214 @@
+    inner: UISpaceServiceRoomList,
+}
+
+impl SpaceServiceRoomList {
+    pub fn new(inner: UISpaceServiceRoomList) -> Self {
+        Self { inner }
+    }
+}
+
+#[matrix_sdk_ffi_macros::export]
+impl SpaceServiceRoomList {
+    pub fn pagination_state(&self) -> SpaceServiceRoomListPaginationState {
+        self.inner.pagination_state().into()
+    }
+
+    pub fn subscribe_to_pagiation_state_updates(
+        &self,
+        listener: Box<dyn SpaceServiceRoomListPaginationStateListener>,
+    ) -> Arc<TaskHandle> {
+        let pagination_state = self.inner.subscribe_to_pagination_state_updates();
+
+        Arc::new(TaskHandle::new(get_runtime_handle().spawn(async move {
+            pin_mut!(pagination_state);
+
+            while let Some(state) = pagination_state.next().await {
+                listener.on_update(state.into());
+            }
+        })))
+    }
+
+    pub fn rooms(&self) -> Vec<SpaceServiceRoom> {
+        self.inner.rooms().into_iter().map(Into::into).collect()
+    }
+
+    pub fn subscribe_to_room_update(
+        &self,
+        listener: Box<dyn SpaceServiceRoomListEntriesListener>,
+    ) -> Arc<TaskHandle> {
+        let entries_stream = self.inner.subscribe_to_room_updates();
+
+        Arc::new(TaskHandle::new(get_runtime_handle().spawn(async move {
+            pin_mut!(entries_stream);
+
+            while let Some(rooms) = entries_stream.next().await {
+                listener.on_update(rooms.into_iter().map(Into::into).collect());
+            }
+        })))
+    }
+
+    pub async fn paginate(&self) -> Result<(), ClientError> {
+        self.inner.paginate().await.map_err(ClientError::from)
+    }
+}
+
+#[derive(uniffi::Enum)]
+pub enum SpaceServiceRoomListPaginationState {
+    Idle { end_reached: bool },
+    Loading,
+}
+
+impl From<UISpaceServiceRoomListPaginationState> for SpaceServiceRoomListPaginationState {
+    fn from(state: UISpaceServiceRoomListPaginationState) -> Self {
+        match state {
+            UISpaceServiceRoomListPaginationState::Idle { end_reached } => {
+                SpaceServiceRoomListPaginationState::Idle { end_reached }
+            }
+            UISpaceServiceRoomListPaginationState::Loading => {
+                SpaceServiceRoomListPaginationState::Loading
+            }
+        }
+    }
+}
+
+#[matrix_sdk_ffi_macros::export(callback_interface)]
+pub trait SpaceServiceRoomListPaginationStateListener:
+    SendOutsideWasm + SyncOutsideWasm + Debug
+{
+    fn on_update(&self, pagination_state: SpaceServiceRoomListPaginationState);
+}
+
+#[matrix_sdk_ffi_macros::export(callback_interface)]
+pub trait SpaceServiceRoomListEntriesListener: SendOutsideWasm + SyncOutsideWasm + Debug {
+    fn on_update(&self, rooms: Vec<SpaceServiceRoom>);
+}
+
+#[matrix_sdk_ffi_macros::export(callback_interface)]
+pub trait SpaceServiceJoinedSpacesListener: SendOutsideWasm + SyncOutsideWasm + Debug {
+    fn on_update(&self, rooms: Vec<SpaceServiceRoom>);
+}
```

Comment:
Up to you but I noticed whilst conforming out `SDKListener` to these that they're identical. So maybe a single generic `SpaceServiceSpaceRoomsListener` could suffice here 🌵

Reply:
Not sure how the cactus got in there, but leaving it because it is cute 😆